docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,242 @@
+sources are bounded and unbounded otherwise.
+
+STREAMING execution mode, on the other hand, can be used for both bounded and
+unbounded jobs.
+
+Another term that is typically used for bounded sources is _batch source_. Or,
+we can say that we are working with a batch data set or a batch of data. Some
+typical other terms for an unbounded source are _continuous source_, _streaming
+source_, _stream_, or _infinite stream_.
+
+As a rule of thumb, you should be using BATCH execution mode when your program
+is bounded because this will be more efficient. You have to use STREAMING
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
+
+TODO: Should we even go into this?
+
+One obvious outlier case is when you want to use a bounded job to bootstrap
+some job state that you then want to use in an unbounded job. For example, by
+running a bounded job using STREAMING mode, taking a savepoint and then
+restoring that savepoint on an unbounded job. This is a very specific use case
+and one that might soon become obsolete when we allow producing a savepoint as
+additional output of a BATCH execution job.
+
+## Configuring BATCH execution mode
+
+The execution mode can be configured via the `execution.runtime-mode` setting.
+There are three possible values:
+
+ - `STREAMING`: The classic DataStream execution mode
+ - `BATCH`: Batch-style execution on the DataStream API
+ - `AUTOMATIC`: Let the system decide based on the boundedness of the sources
+
+ This can be configured either in the `flink-conf.yaml`, via command line
+ parameters of `bin/flink run ...`, or programmatically when
+ creating/configuring the `StreamExecutionEnvironment`.
+
+ Here's how you can configure the execution mode via the command line:
+
+ ```bash
+ $ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
+ ```
+
+ This example shows how you can configure the execution mode in code:
+
+ ```java
+Configuration config = new Configuration();
+config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+ ```
+
+## Execution Behavior
+
+This section provides an overview of the execution behavior of BATCH execution
+mode and contrasts it with STREAMING execution mode. For more details, please
+refer to the FLIPs that introduced this feature:
+[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
+[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
+documentation about [task scheduling (TODO)](<TODO>).
+
+### Task Scheduling And Network Shuffle
+
+Flink jobs consist of different operations that are connected together in a
+dataflow graph. The system decides how to schedule the execution of these
+operations on different processes/machines (TaskManager) and how data is
+shuffled (sent) betwixt them.
```

Comment:
I appreciate learning a new word `betwixt`, but I think it is nicer towards our readers, which are very often not native speakers, to stick to the simpler `between`.

Reply:
I just put it there for someone to find it. 😊

Reply:
I was bemused to encounter such a fine word, and correctly assumed I didn't need to be the one to react to it. 😏

Reply:
🧐